StyleCop.Analyzers/StyleCop.Analyzers.Test/MaintainabilityRules/SA1119UnitTests.cs

```diff
@@ -0,0 +1,1239 @@
+            var testCode = @"public class Foo
+{
+    public void Bar()
+    {
+        var x = from y in new int[10] select y + 1;
+    }
+}";
+            await VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestQueryParenthesis()
+        {
+            var testCode = @"public class Foo
+{
+    public void Bar()
+    {
+        var x = (from y in new int[10] select y + 1);
+    }
+}";
+            var expected = new[]
+            {
+                new DiagnosticResult
+                {
+                    Id = DiagnosticId,
+                    Message = "Statement must not use unnecessary parenthesis",
+                    Severity = DiagnosticSeverity.Warning,
+                    Locations =
+                        new[]
+                        {
+                            new DiagnosticResultLocation("Test0.cs", 5, 17)
+                        }
+                }
+            };
+
+            await VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None);
+
+            var fixedCode = @"public class Foo
+{
+    public void Bar()
+    {
+        var x = from y in new int[10] select y + 1;
+    }
+}";
+            await VerifyCSharpFixAsync(testCode, fixedCode);
+        }
+
+        [TestMethod]
+        public async Task TestQueryInner()
+        {
+            var testCode = @"public class Foo
+{
+    public void Bar()
+    {
+        var x = (from y in new int[10] select y + 1).ToString();
+    }
+}";
+            await VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestBinary()
+        {
+            var testCode = @"public class Foo
+{
+    public void Bar()
+    {
+        var x = 1 + 1 + 1;
+        var y = (1 + 1) + (1 + 1);
+        var z = 1 * ~(1 + 1);
+    }
+}";
+            await VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestInnerTriviaCodeFix()
+        {
+            var testCode = @"public class Foo
+{
+        private bool Test()
+        {
+            string x = """";
+            return (x == null || x.GetType() == typeof(object)
+#if !(NET35 || NET20 || PORTABLE40)
+                    || x.GetType() == typeof(string)
+#endif
+                );
+        }
+}";
+            var fixedCode = @"public class Foo
+{
+        private bool Test()
+        {
+            string x = """";
+            return x == null || x.GetType() == typeof(object)
+#if !(NET35 || NET20 || PORTABLE40)
+                    || x.GetType() == typeof(string)
+#endif
+                ;
+        }
+}";
+            await VerifyCSharpFixAsync(testCode, fixedCode);
+        }
+
+        [TestMethod]
+        public async Task TestCodeFixWorksInTrivia()
+        {
+            var testCode = @"#if (NET20 || NET35)
```

Comment:
I was trying it on Json.net which has a lot of this and had a few issues before. ;)